### PR TITLE
[MPS] LSTM grad_y missing fix

### LINKED ISSUE #96416: Loss.backward() error when using MPS on M1
### 🐛 Describe the bug

I am trying to train a seq2seq model on MPS on an M1 Mac, but during loss.backward() I get a runtime error: `RuntimeError: Expected a proper Tensor but got None (or an undefined Tensor in C++) for argument #0 'grad_y`

```python
# Code to reproduce error
import torch
import torch.nn as nn
import random
from torch.nn import MSELoss


class RNNEncoder(nn.Module):
    def __init__(self, hid_dim=128, n_layers=2, dropout=0.5, device="cpu"):
        super().__init__()

        self.fc = nn.Linear(4, 128)
        self.dropout = nn.Dropout(dropout)

        self.fc_bn = nn.Linear(512, 128)

        self.rnn = nn.LSTM(input_size=128, hidden_size=hid_dim, batch_first=True, bidirectional=False,
                           num_layers=n_layers).to(device)

    def forward(self, x):
        h = self.fc(x)
        _, (hidden, cell) = self.rnn(h)
        return hidden, cell


class RNNDecoder(nn.Module):
    def __init__(self, input_dim=4, n_layers=2, hid_dim=128, dropout=0.5, device="cpu"):
        super().__init__()

        self.rnn = nn.LSTM(input_size=input_dim, hidden_size=hid_dim, num_layers=n_layers)

        self.post_net = nn.Sequential(
            nn.Linear(128, 64),
            nn.ReLU(),
            nn.Dropout(dropout),
            nn.Linear(64, 4)
        )

    def forward(self, dec_input, hidden, cell):
        # Input shape: [batch_size, L, input_size]
        # Reshape to: [L, batch_size, input_size]
        dec_input = dec_input.reshape(dec_input.shape[1], -1, 4)

        rnn_out, (hidden, cell) = self.rnn(dec_input, (hidden, cell))

        # rnn_out = rnn_out.reshape(-1, 128, rnn_out.shape[0])
        rnn_out = rnn_out.reshape(-1, rnn_out.shape[0], 128)
        out = self.post_net(rnn_out)
        return out.flatten(1), (hidden, cell)


class Seq2Seq(nn.Module):
    def __init__(self, encoder, decoder, device="cpu"):
        super().__init__()
        self.encoder = encoder.to(device)
        self.decoder = decoder.to(device)
        self.device = device

    def forward(self, bases, raw, teacher_forcing_ratio=0.5):

        batch_size = raw.shape[0]
        trg_len = raw.shape[1]

        outputs = torch.zeros(batch_size, trg_len).to(self.device)


        hidden, cell = self.encoder(bases)

        # Generate first input of zeros [batch_size, Length 1, input size]
        input_dec = torch.zeros((raw.shape[0], 1, 4)).to(self.device)

        for t in range(0, trg_len, 4):
            output_dec, (hidden, cell) = self.decoder(input_dec, hidden, cell)

            # Check if we are at the end of the target sequence and slice the required prediction length
            if t+4 > trg_len:
                pred_interval = trg_len - t
                outputs[:, t:t+pred_interval] = output_dec[:, :pred_interval]

            else:
                outputs[:, t:t+4] = output_dec

            teacher_force = random.random() < teacher_forcing_ratio

            input_dec = raw[:, t:t+4].unsqueeze(1) if teacher_force else output_dec.unsqueeze(1)

        return outputs


if __name__ == "__main__":
    device = "mps"
    encoder = RNNEncoder(device=device)
    decoder = RNNDecoder(device=device)
    model = Seq2Seq(encoder=encoder, decoder=decoder, device=device)

    loss = MSELoss()

    bases = torch.rand([32, 89, 4]).to(device)
    raw = torch.rand(32, 500).to(device)
    output = model(bases, raw)
    loss_fn = MSELoss()
    loss = loss(output, raw)
    loss.backward()
```

Full error message:
```
Traceback (most recent call last):
  File "/Users/jonas_n/Documents/repositories/NanoGen/src/models/pytorch_forum.py", line 102, in <module>
    loss.backward()
  File "/opt/homebrew/Caskroom/miniforge/base/envs/torch_nightly/lib/python3.10/site-packages/torch/_tensor.py", line 487, in backward
    torch.autograd.backward(
  File "/opt/homebrew/Caskroom/miniforge/base/envs/torch_nightly/lib/python3.10/site-packages/torch/autograd/__init__.py", line 204, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: Expected a proper Tensor but got None (or an undefined Tensor in C++) for argument #0 'grad_y'
```

### Versions

```
PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.2.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:25:29) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.2.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] torch==1.13.1
[pip3] torchaudio==0.13.1
[pip3] torchinfo==1.7.2
[pip3] torchvision==0.14.1
[conda] numpy                     1.24.1          py310h5d7c261_0    conda-forge
[conda] pytorch                   1.13.1                 py3.10_0    pytorch
[conda] torchaudio                0.13.1                py310_cpu    pytorch
[conda] torchinfo                 1.7.2              pyhd8ed1ab_0    conda-forge
[conda] torchvision               0.14.1                py310_cpu    pytorch
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by alexdremov**
Fixes #96416
Added tests that do not use LSTM output simalarly to the issue

Seems like this fix once again introduces backward incompatibility.

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Sound good to me.
Will let @kulinseth take a quick look as well.

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -7219,7 +7219,7 @@
     MPS: _lstm_mps
   autogen: _lstm_mps.out
 
-- func: lstm_mps_backward(Tensor grad_y, Tensor? grad_hy, Tensor? grad_cy, Tensor z_state, Tensor cell_state_fwd, Tensor input, Tensor layersOutputs, Tensor[] hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool batch_first) -> (Tensor, Tensor[], Tensor[])
+- func: lstm_mps_backward(Tensor? grad_y, Tensor? grad_hy, Tensor? grad_cy, Tensor z_state, Tensor cell_state_fwd, Tensor input, Tensor layersOutputs, Tensor[] hx, Tensor[] params, bool has_biases, int num_layers, float dropout, bool train, bool bidirectional, bool batch_first) -> (Tensor, Tensor[], Tensor[])
```
Comment (kulinseth) [inline aten/src/ATen/native/native_functions.yaml:7222]: @albanD , will this affect `compat` tests ? I hope not. Otherwise the change looks good to me.

**Review by kulinseth — APPROVED** [review]

**Comment by albanD** [comment]
It is most likely breaking the compat test. But I'm sure the previous PR added a time so far in the future that it is not over yet haha.
In any case, I don't expect anyone to have serialized this schema. So we can break it. it is fine.

**Pull request closed without merging**